### PR TITLE
Added retries for getting refreshable token

### LINKED ISSUE #321: Retry token refresh for Azure SPN in case of `Temporarily throttled, too many requests`
Symptom: `inner token: token error: {"error":"invalid_request","error_description":"Temporarily throttled, too many requests"}`

Follow-up to https://github.com/databricks/terraform-provider-databricks/issues/2042#issuecomment-1452528306

cc @vmazobhc

### PR DESCRIPTION
**Pull request by tanmay-db**
Fix #321

### CODE REVIEW

File: config/oauth_visitors.go [hunk 1]
```diff
@@ -1,22 +1,38 @@
...
 
+	"github.com/databricks/databricks-sdk-go/retries"
 	"golang.org/x/oauth2"
 )
 
+func retriableTokenSource(ctx context.Context, ts oauth2.TokenSource) (*oauth2.Token, error) {
+	return retries.Poll[oauth2.Token](ctx, 1*time.Minute, func() (*oauth2.Token, *retries.Err) {
+		token, err := ts.Token()
+
+		if strings.Contains(err.Error(), "throttled") {
```
Comment (nfx) [inline config/oauth_visitors.go:18]: a) check if error is not empty.
b) add a list of strings to trigger retries on

File: config/oauth_visitors.go [hunk 2]
```diff
@@ -1,22 +1,38 @@
...
+		return token, nil
+	})
+}
+
 func serviceToServiceVisitor(inner, cloud oauth2.TokenSource, header string) func(r *http.Request) error {
 	refreshableInner := oauth2.ReuseTokenSource(nil, inner)
 	refreshableCloud := oauth2.ReuseTokenSource(nil, cloud)
 	return func(r *http.Request) error {
-		inner, err := refreshableInner.Token()
+		inner, err := retriableTokenSource(context.Background(), refreshableInner)
```
Comment (nfx) [inline config/oauth_visitors.go:30]:
```suggestion
		inner, err := retriableTokenSource(r.Context(), refreshableInner)
```
and in other places as well

**Review by nfx — CHANGES_REQUESTED** [review]
please add tests. e.g. create a mocked token source - e.g. a function type, that has a `Token() (*oauth2.Token, error)` method and test for different retriable and non-retriable scenarios

File: config/oauth_visitors_test.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	"golang.org/x/oauth2"
+)
+
+type mockTokenSource struct {
+	mockedTokenFunc func() (*oauth2.Token, error)
+}
+
+func (m mockTokenSource) Token() (*oauth2.Token, error) {
+	return m.mockedTokenFunc()
+}
```
Comment (tanmay-db) [inline config/oauth_visitors_test.go:18]: It should be outside the tests as it's being used by three tests, putting this inside will increase code duplication.

File: config/oauth_visitors.go [hunk 4]
```diff
@@ -1,22 +1,42 @@
...
 	"golang.org/x/oauth2"
 )
 
+func retriableTokenSource(ctx context.Context, ts oauth2.TokenSource) (*oauth2.Token, error) {
+	return retries.Poll(ctx, 1*time.Minute, func() (*oauth2.Token, *retries.Err) {
+		token, err := ts.Token()
+		if err == nil {
+			return token, nil
+		}
+		var retryKeywords = []string{"throttled", "too many requests", "429", "request limit exceeded", "rate limit"}
```
[the remark this answers is not in the document]
Comment (nfx) [reply config/oauth_visitors.go:20]: this is good enough for now:)

File: config/oauth_visitors_test.go (hunk 3 above)
Comment (nfx) [reply config/oauth_visitors_test.go:18]: it's fine to do it this way.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**